keep/api/routes/preset.py

```diff
@@ -60,6 +64,51 @@
     ),
 }
 
+async def pull_alerts_from_providers(
+    tenant_id: str, pusher_client: Pusher | None, sync: bool = False
+) -> list[AlertDto]:
+    """
+    Pulls alerts from providers and record the to the DB.
+
+    "Get or create logics".
+    """
+
+    context_manager = ContextManager(
+        tenant_id=tenant_id,
+        workflow_id=None,
+    )
+
+    for provider in ProvidersFactory.get_installed_providers(tenant_id=tenant_id):
+        provider_class = ProvidersFactory.get_provider(
+            context_manager=context_manager,
+            provider_id=provider.id,
+            provider_type=provider.type,
+            provider_config=provider.details,
+        )
+        logger.info(
+            f"Pulling alerts from provider {provider.type} ({provider.id})",
+            extra={
+                "provider_type": provider.type,
+                "provider_id": provider.id,
+                "tenant_id": tenant_id,
+            },
+        )
+        sorted_provider_alerts_by_fingerprint = (
+            provider_class.get_alerts_by_fingerprint(tenant_id=tenant_id)
+        )
+        for fingerprint, alert  in sorted_provider_alerts_by_fingerprint.items():
+            await process_event(
+                {},
+                tenant_id,
+                None,
+                None,
+                fingerprint,
+                None,
+                None,
+                alert,
+                save_if_duplicate=False,
+            )
```

Comment:
Another thing here (don't know how important it is but..) is process_event accepts a list of AlertDto, so you can:
```suggestion
        alerts = [alert for alert in sorted_provider_alerts_by_fingerprint.values()]
        await process_event(
            {},
            tenant_id,
            None,
            None,
            fingerprint,
            None,
            None,
            alerts,
            save_if_duplicate=False,
        )
```
or something, but meh :P

Reply:
I didn't want to re-calculate fingerprint there, not sure if it's critical or not 🤔